### PR TITLE
Use Terraform to bcrypt the admin password

### PR DESCRIPTION
**Pull request by edevenport**
Frontend and backend have been adjusted to use Terraform bcrypt interpolation for the adminPassword per issue #643. I have tested locally with `make apply` and through the UI with local `terraform` versions 0.9.11 and 0.10.2.

Password has been tested with the following special characters:

    `-=\][';/.,~!@#$%^&*()_+|}{":?><

Backslashes and double quotes will need to be escaped when setting the password in `terraform.tfvars` manually (template descriptions updated accordingly), but all above special characters handled successfully through the UI.

I am unable to validate tests through Jenkins, but `TF_VAR_tectonic_admin_password_hash` is now `TF_VAR_tectonic_admin_password` and requires a plain text password.

### CODE REVIEW

**Comment by edevenport** [comment]
@squat Do you have some time to review the code and test results?

**Comment by squat** [comment]
Thanks @edevenport, I'll TAL now.

**Comment by sym3tri** [comment]
Jenkins & GitHub are complaining about the same merge conflicts.

**Comment by edevenport** [comment]
ok to test

**Comment by edevenport** [comment]
@sym3tri @squat I merged the latest updates from master and corrected the merge conflicts.

**Comment by cpanato** [comment]
@edevenport there are two small issues in the lint for the frontend.

```
/home/ubuntu/workspace/tectonic-installer_PR-1771-MDYVNSYM5ANLWQBB6KSQTDFNNEI4ELV7WSD4XPYFG4CEHLHS2QWQ/installer/frontend/cluster-config.js
  196:37  error  'opts' is assigned a value but never used  no-unused-vars
  297:43  error  'opts' is assigned a value but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)

```

**Comment by edevenport** [comment]
ok to test

**Comment by edevenport** [comment]
Thank you @cpanato - the lint errors should be resolved now.

**Comment by squat** [comment]
@edevenport looking better! now the frontend tests are failing with:
```
    Difference:
    
    - Expected
    + Received
    
    @@ -9,11 +9,11 @@
       "platform": "aws",
       "pullSecret": "<TECTONIC_PULL_SECRET>",
       "retry": false,
       "variables": Object {
         "tectonic_admin_email": "admin@example.com",
    -    "tectonic_admin_password": "PASSWORD",
    +    "tectonic_admin_password": "pjazbahvw!Singon9cwh^ygsi9hoblxW",
         "tectonic_aws_extra_tags": Object {
           "test_tag": "testing",
         },
         "tectonic_aws_master_custom_subnets": Object {
           "us-west-1a": "10.0.0.0/19",
```
To fix this we need to update the `*.progress` files in `installer/frontend/__tests__/examples/` and replace https://github.com/coreos/tectonic-installer/blob/master/installer/frontend/__tests__/examples/tectonic-aws-vpc.progress#L49 with the new example password you specified in the `*.json` files. Alternatively, the json files could all use `PASSWORD` as the test password.

**Comment by edevenport** [comment]
ok to test

File: Documentation/install/azure/azure-terraform.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
+# Install Tectonic on Azure with Terraform
```
Comment (squat) [inline Documentation/install/azure/azure-terraform.md]: @edevenport why was this included in this PR? could we remove this doc?

**Comment by edevenport** [comment]
Thank you @squat - looks like the tests are passing now.

**Comment by squat** [comment]
@edevenport the tests look good. The one main outstanding issue is the azure doc. Did that sneak in accidentally?

**Comment by edevenport** [comment]
ok to test

File: Documentation/install/azure/azure-terraform.md (hunk 1 above)
Comment (edevenport) [reply Documentation/install/azure/azure-terraform.md]: @squat Sorry about that. It was updated in my original PR before it was removed from master and seems to have gotten restored during my pull. I  rebased and removed the file.

File: config.tf [hunk 2]
```diff
@@ -291,12 +291,12 @@ Note: This field MUST be in all lower-case e-mail address format and set manuall
...
 }
 
-variable "tectonic_admin_password_hash" {
+variable "tectonic_admin_password" {
   type = "string"
 
   description = <<EOF
-The bcrypt hash of admin user password to login to the Tectonic Console.
-Use the bcrypt-hash tool (https://github.com/coreos/bcrypt-tool/releases/tag/v1.0.0) to generate it.
+The admin user password to login to the Tectonic Console. Backslashes and double quotes should
```
Comment (squat) [inline config.tf]: s/should/must/ to convey that it is a requirement.

File: modules/tectonic/variables.tf [hunk 3]
```diff
@@ -65,10 +65,10 @@ variable "admin_email" {
   default     = "admin@example.com"
 }
 
-variable "admin_password_hash" {
-  description = "Hashed password used to by the cluster admin to login to the Tectonic Console. Generate with the bcrypt-hash tool (https://github.com/coreos/bcrypt-tool/releases/tag/v1.0.0)."
+variable "admin_password" {
+  description = "Password used to by the cluster admin to login to the Tectonic Console. Backslashes and double quotes should be escaped."
```
Comment (squat) [inline modules/tectonic/variables.tf]: s/should/must/ to convey that it is a requirement.

File: Documentation/variables/config.md [hunk 4]
```diff
@@ -7,7 +7,7 @@ This document gives an overview of variables used in all platforms of the Tecton
 | Name | Description | Type | Default |
 |------|-------------|:----:|:-----:|
 | tectonic_admin_email | The e-mail address used to: 1. login as the admin user to the Tectonic Console. 2. generate DNS zones for some providers.<br><br>Note: This field MUST be in all lower-case e-mail address format and set manually prior to creating the cluster. | string | - |
-| tectonic_admin_password_hash | The bcrypt hash of admin user password to login to the Tectonic Console. Use the bcrypt-hash tool (https://github.com/coreos/bcrypt-tool/releases/tag/v1.0.0) to generate it.<br><br>Note: This field MUST be set manually prior to creating the cluster. | string | - |
+| tectonic_admin_password | The admin user password to login to the Tectonic Console. Backslashes and double quotes should be escaped.<br><br>Note: This field MUST be set manually prior to creating the cluster. | string | - |
```
Comment (squat) [inline Documentation/variables/config.md]: Thanks for making that change. Looks like we need to `make docs && make examples` again now

File: config.tf (hunk 2 above)
Comment (edevenport) [reply config.tf]: Done. I also moved the special character requirement to the `Note` line for better division between parameter description and requirements.

File: modules/tectonic/variables.tf (hunk 3 above)
Comment (edevenport) [reply modules/tectonic/variables.tf]: Done.

File: Documentation/variables/config.md (hunk 4 above)
Comment (edevenport) [reply Documentation/variables/config.md]: And done.

**Comment by squat** [comment]
The tests are failing with the following error:
```
1 error(s) occurred:

* module.tectonic.template_dir.tectonic: 1 error(s) occurred:

* module.tectonic.template_dir.tectonic: 1:3: unknown function called: bcrypt in:

${bcrypt(var.admin_password, 12)}
Makefile:44: recipe for target 'apply' failed
```
This is because the smoke tests use Terraform 0.9.x, which does not have support for bcrypt. We'll need to bump that or wait for #1841.

**Comment by cpanato** [comment]
@edevenport do you mind to rebase this PR? Thanks!

**Comment by cpanato** [comment]
#1841 is merged @squat

**Comment by squat** [comment]
@cpanato @edevenport great now this pr can move forward!

File: platforms/azure/tectonic.tf [hunk 5]
```diff
@@ -55,8 +55,8 @@ module "tectonic" {
   license_path     = "${var.tectonic_vanilla_k8s ? "/dev/null" : pathexpand(var.tectonic_license_path)}"
   pull_secret_path = "${var.tectonic_vanilla_k8s ? "/dev/null" : pathexpand(var.tectonic_pull_secret_path)}"
 
-  admin_email         = "${var.tectonic_admin_email}"
-  admin_password_hash = "${var.tectonic_admin_password_hash}"
+  admin_email    = "${var.tectonic_admin_email}"
+  admin_password = "${var.tectonic_admin_password}"
```
Comment (s-urbaniak) [inline platforms/azure/tectonic.tf]: general concern from my side: this could end up in the tfstate file in cleartext. we have to make sure this is not the case.

Comment (edevenport) [reply platforms/azure/tectonic.tf]: @s-urbaniak I checked the generated `tfstate` file after running a test and the encrypted value is being stored:

```"vars.admin_password_hash": "$2a$12$TrRQ6hcDeE/ydq49r9.z4.YNjrIncDia3ojSJa7aXbktFe40jdH6O",```

So it appears the `modules/tectonic/assets.tf` is storing the encrypted value in the `tfstate` rather than the clear-text value.

**Comment by edevenport** [comment]
A merge conflict was introduced recently, so I rebased the branch and corrected the issue.

**Comment by edevenport** [comment]
@squat We are ready for a review (and hopefully merger) again.

File: tests/smoke/bare-metal/vars/metal.tfvars [hunk 6]
```diff
@@ -113,9 +113,8 @@ tectonic_ssh_authorized_key = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDPQFdwVLr+a
 
 # Tectonic
 
-# generate a hash with the bcrypt-hash tool (https://github.com/coreos/bcrypt-tool/releases/tag/v1.0.0) that will be used for your admin user.
 tectonic_admin_email = "admin@example.com"
-tectonic_admin_password_hash = "$2a$12$T8hTe.NlOPDP0SS3DxNeDuVhHSFbdGXZEhGps/W.BG4QC7.1/nDaG"
+tectonic_admin_password = "password"
```
Comment (s-urbaniak) [inline tests/smoke/bare-metal/vars/metal.tfvars]: Let's get rid of this here and inject it via an env var in CI, @cpanato do you mind to give directions here?

File: modules/tectonic/variables.tf [hunk 7]
```diff
@@ -60,10 +60,10 @@ variable "admin_email" {
   default     = "admin@example.com"
 }
 
-variable "admin_password_hash" {
-  description = "Hashed password used to by the cluster admin to login to the Tectonic Console. Generate with the bcrypt-hash tool (https://github.com/coreos/bcrypt-tool/releases/tag/v1.0.0)."
+variable "admin_password" {
+  description = "Password used to by the cluster admin to login to the Tectonic Console. Backslashes and double quotes must be escaped."
   type        = "string"
-  default     = "2a$12$k9wa31uE/4uD9aVtT/vNtOZwxXyEJ/9DwXXEYB/eUpb9fvEPsH/kO"
+  default     = "password"
```
Comment (s-urbaniak) [inline modules/tectonic/variables.tf]: I strongly suggest to remove this default value here.

File: installer/frontend/__tests__/examples/tectonic-baremetal.progress [hunk 8]
```diff
@@ -37,7 +37,7 @@
     "tectonicLicense": "<TECTONIC_LICENSE>",
     "pullSecret": "<TECTONIC_PULL_SECRET>",
     "adminEmail": "admin@example.com",
-    "adminPassword": "PASSWORD",
+    "adminPassword": "password",
```
Comment (s-urbaniak) [inline installer/frontend/__tests__/examples/tectonic-baremetal.progress]: those variables should be dependency-injected via CI, /cc @cpanato

**Comment by s-urbaniak** [comment]
ping @edevenport: do you mind to rebase+address the review comments?

File: tests/smoke/bare-metal/vars/metal.tfvars (hunk 6 above)
Comment (edevenport) [reply tests/smoke/bare-metal/vars/metal.tfvars]: @cpanato I have no visibility into Jenkins. Would you be able to assist me here?

File: installer/frontend/__tests__/examples/tectonic-baremetal.progress (hunk 8 above)
Comment (edevenport) [reply installer/frontend/__tests__/examples/tectonic-baremetal.progress]: @cpanato Same issue here with no visibility into Jenkins.

**Comment by edevenport** [comment]
@s-urbaniak The rebase is completed. I'll see if @cpanato can assist with the Jenkins piece..

**Comment by alexsomesan** [comment]
retest this please

File: Jenkinsfile [hunk 9]
```diff
@@ -21,6 +21,12 @@ def creds = [
     variable: 'TF_VAR_tectonic_azure_ssh_key'
   ],
   [
+    $class: 'UsernamePasswordMultiBinding',
+    credentialsId: 'tectonic-console-login',
+    passwordVariable: 'TF_VAR_tectonic_admin_email',
+    usernameVariable: 'TF_VAR_tectonic_admin_password'
```
Comment (cpanato) [inline Jenkinsfile]: We dont need to set the admin_email/password because in the rspec tests we randon generate those.

Comment (cpanato) [reply Jenkinsfile]: can you please remove this change?

**Comment by cpanato** [comment]
@edevenport @alexsomesan @s-urbaniak all green now :)
